### ISSUE TITLE
Debug "step" and  "resume" operations are grayed out - 2023-12 M2 + Java 21.0.1

### ISSUE THREAD

**Issue by minduch** [eclipse-jdt/eclipse.jdt.debug#345]
Eclipse 2023-12 M2, Java Temurin 21.0.1.
Microsoft Windows [Version 10.0.23580.1000]

Launching debug of a Java Application project (uses Virtual Threads in case it matters) hits a breakpoint during its startup (after running a couple of seconds) the breakpoint `System.out.println("HERE!");` is encountered. Code is compiled for Java 17.

![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/28196566/07b6e03b-ad0a-47a6-bf46-5e9c946b8326)

The debug operations "step" (into/over/return) and "resume" are grayed out and cannot be performed. Also note the tree view having two empty lines. It doesn't matter which thread I select in the Debug view, the operations in question are always disabled:

![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/28196566/6be3f01e-c2fe-4c7a-b05c-0e4670084c5c)

This problem can be reproduced all the time.

**Comment by iloveeclipse** [comment]
1) Is this a regression, and if yes, which version works?
2) Any chance to provide steps to reproduce / example project?

**Comment by minduch** [comment]
Worse in 2023-09 GA:

![image](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/28196566/e85797b9-1f73-4a46-b91b-fddc7f00a011)

However, with Java Temurin - Eclipse Adoptium 17.0.9+9 - it works fine in both IDE's. So it must be Java 17/21 related, here 2023-09 GA:

![image](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/28196566/5f45e807-f347-442f-af21-b9ed5fd192fc)

and also works in 2023-12 M2:

![image](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/28196566/f021435a-ac95-47f5-9744-755e8d8791f5)

Must I try to find a sample project, I don't have many of those?

**Comment by iloveeclipse** [comment]
I'm lost. Ist this virtual threads related, how it can work on Java 17 which does not have virtual threads? Is this not related to virtual threads, what is the key difference in what you are doing & the rest of the world with working debugger?

Please try to isolate the root cause of your problem. IDE update, Java update, specific code you have or whatever else. 

Once you have that, try to provide steps to reproduce for us so we can fix the problem, whatever it is.

**Comment by minduch** [comment]
I think it's Java 21-related, or to be more sure, post-Java 17 specific. Virtual Threads were implemented from Java 19 and better, and we have preview support for it, but it's disabled. So the code being used targets Java 21 (only for the Virtual Threads part).

It's not IDE. It's Java 21. Using Java 20 (OpenJDK 64-Bit Server VM Temurin-20.0.2+9 (build 20.0.2+9, mixed mode, sharing)) and "--enable-preview" (VirtualThreads are used), the break-point enables the debug operations (step/resume), but has a problem in the Debug view (one thread missing that is sitting on the same breakpoint too, but in another thread):

![image](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/28196566/738bb380-5f82-47cb-9ff7-20b2617979a1)

When stepping and debugging code, the Debug view thread that is "missing":

![image](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/28196566/26f9b5af-4279-48f4-be96-6965ea4b7ae9)

Turning off "--enable-preview" in Java 20 makes it work:

![image](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/28196566/e4dd141c-b381-4c22-a706-5672a7aa2efb)

We have a command line option to disable Virtual Threads, so this problem seems to become clear now.
The Debug view with Virtual Threads has a problem when it hits breakpoints in one or more threads. This is using Java 21 without "--enable-preview" and our VirtualThreads turned OFF:

![image](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/28196566/50003036-064c-41ea-9578-e5d111c93405)

The interesting thing is that we only have one breakpoint at a time when VirtualThreads are off. So I guess somebody with more experience in this area should look at the problem (as it involves two "probably" Virtual Threads being break-point'ed, but I'm not sure).

**Comment by minduch** [comment]
@iloveeclipse Yep, I thinks this is an accurate description from the tests I have performed.

**Comment by iloveeclipse** [comment]
OK, could you please try to isolate some simple example that would allow us to reproduce the issue?
Or, if the application is available, provide steps to reproduce?

If this is caused by virtual threads, I assume one precondition is a breakpoint in code executed by a virtual thread. Most likely you have multiple virtual threads running in one "real" thread and hitting same breakpoint one after each other, and debugger confused by that (because it would be same real thread)?

**Comment by iloveeclipse** [comment]
I wonder if you can change the breakpoint to print current thread name, what you would see on console?
Also assuming there are multiple hits on same breakpoint, could you set hit count to one, just to see if debugger would "work" again? Unfortunately I have not much time to play with that, but I guess you could try to compose  simple example with virtual threads hitting same breakpoint again and again. If you could do that & it will reproduce the issue, it would help a lot.

**Comment by minduch** [comment]
Well, that was not much help. If I use System.err, it will probably sync the threads somehow so then I don't get the problem at all. I however changed the code a bit instead of `System.err.println("HERE!");` to
```java
      ByteArrayOutputStream out=new ByteArrayOutputStream();
      try(PrintStream ps=new PrintStream(out))
        {
        new Exception("StackTrace").printStackTrace(ps);
        }
      System.err.println("HERE: "+new String(out.toByteArray()));
```
And set the break-point on the `System.err` line, and now THAT thread is OK in Debug view (display + enabled step/resume operations), and the other thread is NOT:
![image](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/28196566/77097150-0860-4e95-ba57-5255b7550af8)
Above, I highlighted the other thread not shown in the Debug view, so the step/resume operations are disabled.

I'm not becoming less confused releasing all threads and looking at the printed stack traces, they all come from the same thread as it looks like, and there is no point in sharing this info here - it's just a plain stacktrace from the two `System.err` calls, all originating from `lThread[#48,stderr]/runnable@ForkJoinPool-1-worker-2  HERE: java.lang.Exception: StackTrace`, the name `lThread` indicates it's a Virtual Thread I think.

**Comment by fdummert** [comment]
We're having a multi-threaded app based on virtual threads. We're currently using Eclipse 20231201-2043 with jdt debug 3.21.200.v20231103-0755. The described behavior happens indeed when there are two suspended virtual threads. For me, it displays fine for a moment, but then after a moment one suspended thread disappears from the debug view, leaving blank lines:
![Screenshot_20240229_102320](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/3638091/17ab3d37-b176-4ca4-884b-d2e0f208e61c)
But the most annoying issue is, even with only one virtual thread on a breakpoint, that after <10s, with or without any action like stepping, expanding variables, ... the debug view collapses, variable view gets closed, stepping is disabled, and we have again to open the suspended thread, see screencast:
https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/3638091/b6426f63-cff2-446a-b429-8d136a9f7201
Sometimes the suspended virtual thread is even not shown at all:
![Screenshot_20240229_102342](https://github.com/eclipse-jdt/eclipse.jdt.debug/assets/3638091/86db36cc-7012-4f1e-ba17-e37afe90cbac)
That makes debugging virtual threads really painful. I'm not sure if this is easily reproducible with a toy app, but this is happening in many different apps we have, all based on virtual threads. Maybe for a reproducer you need to setup a simple web server app based on virtual threads (for example Jetty with workers based on virtual threads) and try to debug incoming requests to your test app.

**Comment by iloveeclipse** [comment]
@fdummert, @minduch : if you would provide a simple self containing example project it would increase a chance for this bug to be addressed at some time.

**Comment by fdummert** [comment]
Hi @iloveeclipse, please use https://github.com/fdummert/EclipseVThreads.git as an example project that demonstrates the faulty behaviors. All three debugging blockers can be seen there: collapse of the debug view, vanishing of the suspended thread in the debug view, display of an empty line instead of the suspended thread in the debug view. A few details on how to debug are described in the readme file.

**Comment by christian-elsen** [comment]
The problem still exists. Is there any further info required for to continue on this task? At time of writing the issue is still labeled with 'needinfo'.

**Comment by jukzi** [comment]
Can anyone interested in Virtual threads please test https://github.com/eclipse-jdt/eclipse.jdt.debug/pull/535 ?
@christian-elsen @minduch @fdummert

**Comment by fdummert** [comment]
@jukzi: We did a couple of tests  with #535 and with this PR appears to be working fine!

**Comment by jukzi** [comment]
Is this still an issue or fix so that it can it be closed?

**Comment by fdummert** [comment]
@jukzi: as per my [comment](https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/345#issuecomment-2426900243) above, it is working flawlessly for us

**Comment by minduch** [comment]
I haven’t seen this problem for a while. For me, you can close it. We can always create another ticket in case it comes back.
